### PR TITLE
[Internal Observability] [Part 1] Centralize existing metrics to metric_defs.h

### PR DESCRIPTION
**Pull request by rkooo567**
## Why are these changes needed?

This PR centralizes all existing metrics to `metric_defs.h`. 

Previously, each file relies on implicit import of metric_def.h within the stats module. After this PR we only precisely import `metric_defs.h` for each file.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: src/ray/raylet/node_manager.cc
Comment (SongGuyang) [inline src/ray/raylet/node_manager.cc:2201]: A little weird. Here you don't need to get the debug string.

Comment (SongGuyang) [reply src/ray/raylet/node_manager.cc:2201]: Maybe the debug function could be like:
```
std::string RecordStats(bool return_debug_string=true, bool record_metrics=true);
```
?

Comment (rkooo567) [reply src/ray/raylet/node_manager.cc:2201]: I think RecordStats is also the wrong name what about

```
std::string DebugString(bool record=true) const;
```
? 

Btw, why do we not need empty metrics here? Can we just remove this logic?

Comment (SongGuyang) [reply src/ray/raylet/node_manager.cc:2201]: It is ok If you always need a debug string.
I don't know. I'm not familiar with this part.

Comment (rkooo567) [reply src/ray/raylet/node_manager.cc:2201]: Hmm I think we need some clean interface here. I will discuss with @scv119 and get back here.

File: src/ray/object_manager/object_manager.cc
Comment (ericl) [inline src/ray/object_manager/object_manager.cc:29]: Shouldn't the errors be labels on num_chunks_received, instead of defining separate stats?

File: src/ray/object_manager/object_manager.cc
Comment (ericl) [inline src/ray/object_manager/object_manager.cc:860]: Maybe take a `bool record_metrics = false`?

File: src/ray/object_manager/plasma/stats_collector.cc
Comment (ericl) [inline src/ray/object_manager/plasma/stats_collector.cc:26]: Descriptions seem wrong?

File: src/ray/object_manager/plasma/store.cc
Comment (ericl) [inline src/ray/object_manager/plasma/store.cc:63]: Could we put all the stats defs in a common file? Then we can put an OWNERS file on it to enforce stats standards.

File: src/ray/object_manager/plasma/store.cc
Comment (ericl) [inline src/ray/object_manager/plasma/store.cc:60]: The naming here is not sufficient to understand what this stats is for.

**Review by ericl — CHANGES_REQUESTED** [review]
A couple high level asks:
1. Can we pull out the stat defs into a single file to centralize stat definitions and require CODEOWNERS approval? I've seen this work well in the past, where gatekeeping stats keeps their quality high.
2. It seems weird to combine debug string and reporting metrics. If the overhead isn't large, why not keep it separate?

**Comment by rkooo567** [comment]
> It seems weird to combine debug string and reporting metrics. If the overhead isn't large, why not keep it separate?

I think the concern is overhead + duplicated code (since some debug state requires to go through some data structure). I think we can probably make each component to have Stats struct to solve this. Let me try that out

File: src/ray/object_manager/plasma/store.cc
Comment (rkooo567) [reply src/ray/object_manager/plasma/store.cc]: This required a bit of refactoring changes, but I think it is working now.

**Comment by rkooo567** [comment]
I think we can consolidate some metrics with labels. Should we do it in this PR?

File: src/ray/object_manager/object_manager.cc
Comment (ericl) [inline src/ray/object_manager/object_manager.cc:886]: Since these are new stats, can we consolidate them under a single `object_manager_chunks_received` metric with labels for breakdown?

File: src/ray/object_manager/plasma/stats_collector.cc
Comment (ericl) [inline src/ray/object_manager/plasma/stats_collector.cc:172]: How about

plasma_num_local_objects
plasma_num_local_bytes

File: src/ray/object_manager/plasma/stats_collector.cc
Comment (ericl) [inline src/ray/object_manager/plasma/stats_collector.cc:182]: We shouldn't have "by type" and "by state". Instead, you can label a single metric with both labels, like

ray::stats::STATS_num_objects.Record(num_bytes_foo, {{"Type", "foo"}, {"State", "bar"}}).

Note you might have to refactor the recording a bit.

File: src/ray/object_manager/pull_manager.cc
Comment (ericl) [inline src/ray/object_manager/pull_manager.cc:748]: Should use labels for breakdown. Also, please prefix metrics with compoment, like `STATS_pull_manager_`

File: src/ray/object_manager/push_manager.cc
Comment (ericl) [inline src/ray/object_manager/push_manager.cc:81]:
```suggestion
  ray::stats::STATS_num_chunks_remaining.Record(NumChunksRemaining());
```

Also, should consolidate and use labels for breakdown.

**Review by ericl — COMMENTED** [review]
@rkooo567 how about breaking this PR into a few pieces? There's a lot to review.

1. Refactoring existing metrics to metrics_def.h (should be a quick review).
2. New metric def proposals (review to get right naming and label breakdowns).
3. The rest changing debug string / etc. to record the metrics?

**Comment by rkooo567** [comment]
@ericl Do you mind combining 1 and 3? I think they are less controversial and easier to break down with that

**Comment by ericl** [comment]
You can't do (3) without (2) right? Let's check in the metric definitions first before implementing their recordings, since the defs are where we should be thinking hard about design.

**Comment by rkooo567** [comment]
> You can't do (3) without (2) right? Let's check in the metric definitions first before implementing their recordings, since the defs are where we should be thinking hard about design.

We can just leave the function empty if there's no metric? I don't think we will change how to record metrics (it just calls the children components' RecordMetrics). Combining (1) and (3) will help me splitting the PR a lot nicely & quickly.

**Comment by rkooo567** [comment]
Let me try (1)+(3) first, and if you don't like it I will go with 1->2->3 approach!

**Comment by ericl** [comment]
Sure, let's see how it looks.

**Comment by rkooo567** [comment]
Can you take a look? It is a bit big, but most of code is just moving things around. The only files you need to pay attention is `clsuter_task_manager.cc` and `push_manager.cc`

File: src/ray/object_manager/object_manager.cc
Comment (ericl) [inline src/ray/object_manager/object_manager.cc:851]: These changes are unrelated and could be split into a separate PR right?

File: src/ray/object_manager/plasma/store.cc
Comment (ericl) [inline src/ray/object_manager/plasma/store.cc:499]: This seems to be a no-op change.

File: src/ray/object_manager/plasma/store.cc
Comment (ericl) [inline src/ray/object_manager/plasma/store.cc:108]: Why did the format change here?

**Review by ericl — CHANGES_REQUESTED** [review]
@rkooo567 this is still a pretty large PR. It would be great to split the trivial changes out into another PR that can be easily stamped, so the nontrivial changes can be more carefully reviewed. As is, it's hard to iterate on multiple rounds of reviews due to the large number of files.

(also, should make sure to update the title / description).

**Comment by rkooo567** [comment]
In progress with splitting out some changes...

File: src/ray/object_manager/plasma/store.cc
Comment (rkooo567) [reply src/ray/object_manager/plasma/store.cc:499]: I think it is changed by lint?

**Comment by rkooo567** [comment]
Separated some changes. I really don't think the current PR has any complexity that requires careful review, but if you still want me to separate out RecordMetrics logic from the PR, I can do that.

**Comment by ericl** [comment]
> Separated some changes that you requested. (PR description also updated) I don't think the current PR has complexity that requires careful review, but if you don't agree with me, I can separate logics for RecordMetrics logic from the PR. I think if we separate out that into a different PR, we might have around 300~ish lines of code change.

Please separate it out into another PR. For any PR >250 lines, we should be in the habit of trying to break them into logical components. There's not really a need to mix them.

File: src/ray/object_manager/plasma/store.cc
Comment (rkooo567) [inline src/ray/object_manager/plasma/store.cc:534]: It was done by lint. I don't know why this happened.

**Comment by rkooo567** [comment]
This only contains changes that centralize the metric definition to a single file now.

File: src/ray/stats/metric_defs.cc
Comment (ericl) [inline src/ray/stats/metric_defs.cc:29]:
```suggestion
  Syntax sugar to define a metrics:
```

**Review by ericl — APPROVED** [review]

**Comment by ericl** [comment]
LGTM. We'll have to audit these, but that can come later.

**Comment by rkooo567** [comment]
I think we can comment out the code owner stuff after we merge the PRs!

File: src/ray/stats/metric_defs.h
Comment (scv119) [inline src/ray/stats/metric_defs.h:19]: super nit: separate MACRO from the stats definition.

File: src/ray/stats/metric_defs.h
Comment (scv119) [inline src/ray/stats/metric_defs.h:218]: this comment could go away?

**Review by scv119 — APPROVED** [review]

File: src/ray/stats/metric_defs.h
Comment (rkooo567) [reply src/ray/stats/metric_defs.h:218]: I will remove this once we migrate all metrics to DEFINE_stats!

**Pull request merged**
